src/Features/Core/Diagnostics/EngineV1/DiagnosticIncrementalAnalyzer_BuildSynchronization.cs

```diff
@@ -0,0 +1,192 @@
+        }
+
+        private static ImmutableArray<DiagnosticData> GetExistingDiagnostics(AnalysisData analysisData)
+        {
+            if (analysisData == null)
+            {
+                return ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            return analysisData.Items;
+        }
+
+        private ImmutableArray<DiagnosticData> MergeDiagnostics(ImmutableArray<DiagnosticData> liveDiagnostics, ImmutableArray<DiagnosticData> existingDiagnostics)
+        {
+            ImmutableArray<DiagnosticData>.Builder builder = null;
+
+            if (liveDiagnostics.Length > 0)
+            {
+                builder = ImmutableArray.CreateBuilder<DiagnosticData>();
+                builder.AddRange(liveDiagnostics);
+            }
+
+            if (existingDiagnostics.Length > 0)
+            {
+                builder = builder ?? ImmutableArray.CreateBuilder<DiagnosticData>();
+                builder.AddRange(existingDiagnostics.Where(d => d.Severity == DiagnosticSeverity.Hidden));
+            }
+
+            return builder == null ? ImmutableArray<DiagnosticData>.Empty : builder.ToImmutable();
+        }
+
+        private static ILookup<string, DiagnosticData> CreateDiagnosticIdLookup(ImmutableArray<DiagnosticData> diagnostics)
+        {
+            if (diagnostics.Length == 0)
+            {
+                return null;
+            }
+
+            return diagnostics.ToLookup(d => d.Id);
+        }
+
+        private ImmutableArray<DiagnosticData> ConvertToLiveDiagnostics(
+            ILookup<string, DiagnosticData> lookup, ImmutableArray<DiagnosticDescriptor> descriptors, HashSet<string> seen)
+        {
+            if (lookup == null)
+            {
+                return ImmutableArray<DiagnosticData>.Empty;
+            }
+
+            ImmutableArray<DiagnosticData>.Builder builder = null;
+            foreach (var descriptor in descriptors)
+            {
+                // make sure we don't report same id to multiple different analyzers
+                if (seen.Contains(descriptor.Id))
+                {
+                    // TODO: once we have information where diagnostic came from, we probably doesnt need this.
+                    continue;
+                }
+
+                var items = lookup[descriptor.Id];
+                if (items == null)
+                {
+                    continue;
+                }
+
+                // record that we processed this.
+                seen.Add(descriptor.Id);
+
+                builder = builder ?? ImmutableArray.CreateBuilder<DiagnosticData>();
+                builder.AddRange(items.Select(d => CreateLiveDiagnostic(descriptor, d)));
+            }
+
+            return builder == null ? ImmutableArray<DiagnosticData>.Empty : builder.ToImmutable();
+        }
+
+        private static DiagnosticData CreateLiveDiagnostic(DiagnosticDescriptor descriptor, DiagnosticData diagnostic)
```

Comment:
So we are embedding the descriptor fields within the build diagnostic? What does the build diagnostic originally contain for these fields such as Id, Category, etc.?